### PR TITLE
[C++ API] Add functional overloads for activation, batchnorm, distance, embedding

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29360 Add functional overloads for fold, linear, loss, normalization, padding
* #29359 Add functional overloads for pixelshuffle, pooling, upsampling, vision
* **#29358 Add functional overloads for activation, batchnorm, distance, embedding**

This PR adds functional overloads that take the full set of arguments (instead of just Options) for the following functionals:
- activation
- batchnorm
- distance
- embedding

These new functionals lives in the `torch::nn::functional::detail` namespace and they are only meant to be called from the module forward methods (i.e. they are not public API). This is in preparation for the future change where we make module Options and functional Options two different classes, because if the module forward method has to construct a new functional Options object every time it runs it will be pretty silly and bad performance.

Differential Revision: [D18376976](https://our.internmc.facebook.com/intern/diff/D18376976)

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 2b21de5:
* 0/1 recognized as flaky
* 1/1 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=2b21de5001249396512720afc3c3b261303a0d66).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Review by pbelevich — APPROVED** [review]
